File: src/desktop/ax-role-map.ts
/**
 * AXRole → unified role 映射表（parse4 §4.3）
 *
 * 与 rust-helper/src/ax_role_map.rs **field-by-field 镜像**。
 * 同一个 AXRole 字符串在两端必须映射到相同的 unified role。
 *
 * 维护规则（铁律）：
 *  1. Rust 端增改 → TS 端同步增改（反之亦然）
 *  2. 覆盖率验证：M0.5a 采 10 个 native app 各 3 window 验证 ≥80%
 *     （13 §3.4 M0.5a 第 3 条）
 *  3. fallback：未在表内的 AXRole 一律 → "unknown"
 *
 * INV-21 注：本表 key 是 AXRole 字符串值（数据值，从 AXAPI 流入），
 * 不是平台 API 名（如 AXUIElement / AXPress）。这些 AX* 字符串是苹果
 * 已公开的 role 标识符（数据契约），不是 API 调用符号。为可读性 + 与
 * Rust 端 1:1 对账可读，本表保留 "AX" 前缀；它们是 wire 上的数据字典。
 * 真正隔离在 Rust 的是 AXUIElement* / CGEvent* / AXPress 等 API 调用符号。
 *
 * 借鉴：13 §4.3；WAI-ARIA role 标准（DOM-like 命名）；08 附录 A。
 */

// ============================================================
// 映射表（与 ax_role_map.rs match 项 1:1）
// ============================================================
export const AX_ROLE_MAP: Record<string, string> = {
  // Buttons / actions
  AXButton: "button",
  AXPopUpButton: "select",
  AXCheckBox: "checkbox",
  AXRadioButton: "radio",
  AXMenuButton: "menubutton",
  // Text
  AXTextField: "textfield",
  AXTextArea: "textarea",
  AXStaticText: "text",
  // Menus
  AXMenu: "menu",
  AXMenuItem: "menuitem",
  AXMenuBar: "menubar",
  AXMenuBarItem: "menubaritem",
  // Windows / sheets
  AXWindow: "window",
  AXSheet: "dialog", // modal sheet
  AXPopover: "popover",
  // Lists / trees / tables
  AXRow: "row",
  AXOutline: "tree",
  AXList: "list",
  AXTable: "table",
  // Layout / groups
  AXScrollArea: "scrollarea",
  AXTabGroup: "tablist",
  AXToolbar: "toolbar",
  AXGroup: "group",
  AXLayoutArea: "group", // Xcode storyboard canvas
  // Media
  AXImage: "img",
  AXLink: "link",
  // Generic / unknown
  AXUnknown: "unknown",
  // Generic 应用级角色兜底（非 DOM 角色；保留原值便于 doctor 报告）
  AXApplication: "application",
  AXSystemWide: "systemwide",
};

/**
 * AXRole 字符串 → unified role。未在表内的 AXRole 一律 → "unknown"。
 *
 * 与 Rust 端 `map_ax_role(ax: &str) -> &'static str` 同语义。
 */
export function mapAxRole(ax: string): string {
  return AX_ROLE_MAP[ax] ?? "unknown";
}

File: rust-helper/tests/ax_role_map.rs
//! AXRole 映射全覆盖测试（平台无关，CI 在 Linux 也跑）。
//!
//! parse4 §5.1 ax_role_map.rs + §4.3 映射表。
//! 任何映射变更必须同步改 src/ax_role_map.rs + src/desktop/ax-role-map.ts（TS 端）。

use lasso_rust_helper::ax_role_map::map_ax_role;

#[test]
fn buttons() {
    assert_eq!(map_ax_role("AXButton"), "button");
    assert_eq!(map_ax_role("AXPopUpButton"), "select");
    assert_eq!(map_ax_role("AXCheckBox"), "checkbox");
    assert_eq!(map_ax_role("AXRadioButton"), "radio");
    assert_eq!(map_ax_role("AXMenuButton"), "menubutton");
}

#[test]
fn text_inputs() {
    assert_eq!(map_ax_role("AXTextField"), "textfield");
    assert_eq!(map_ax_role("AXTextArea"), "textarea");
    assert_eq!(map_ax_role("AXStaticText"), "text");
}

#[test]
fn menus() {
    assert_eq!(map_ax_role("AXMenu"), "menu");
    assert_eq!(map_ax_role("AXMenuItem"), "menuitem");
    assert_eq!(map_ax_role("AXMenuBar"), "menubar");
    assert_eq!(map_ax_role("AXMenuBarItem"), "menubaritem");
}

#[test]
fn windows_and_dialogs() {
    assert_eq!(map_ax_role("AXWindow"), "window");
    assert_eq!(map_ax_role("AXSheet"), "dialog"); // modal sheet
    assert_eq!(map_ax_role("AXPopover"), "popover");
}

#[test]
fn lists_trees_tables() {
    assert_eq!(map_ax_role("AXRow"), "row");
    assert_eq!(map_ax_role("AXOutline"), "tree");
    assert_eq!(map_ax_role("AXList"), "list");
    assert_eq!(map_ax_role("AXTable"), "table");
}

#[test]
fn layout_groups() {
    assert_eq!(map_ax_role("AXScrollArea"), "scrollarea");
    assert_eq!(map_ax_role("AXTabGroup"), "tablist");
    assert_eq!(map_ax_role("AXToolbar"), "toolbar");
    assert_eq!(map_ax_role("AXGroup"), "group");
    assert_eq!(map_ax_role("AXLayoutArea"), "group"); // Xcode storyboard canvas
}

#[test]
fn media() {
    assert_eq!(map_ax_role("AXImage"), "img");
    assert_eq!(map_ax_role("AXLink"), "link");
}

#[test]
fn unknown_falls_back() {
    assert_eq!(map_ax_role("AXUnknown"), "unknown");
    assert_eq!(map_ax_role("AXSomeFutureRole"), "unknown");
    assert_eq!(map_ax_role(""), "unknown");
}

#[test]
fn application_root_roles_kept_for_doctor() {
    // 不映射到 DOM 角色；doctor / 调试输出保留
    assert_eq!(map_ax_role("AXApplication"), "application");
    assert_eq!(map_ax_role("AXSystemWide"), "systemwide");
}

#[test]
fn consistency_check_returns_static_str() {
    // 映射返回 &'static str — 调用方安全存储无需 clone
    let r: &'static str = map_ax_role("AXButton");
    assert_eq!(r, "button");
}
